### ISSUE TITLE
Add ADABOUND Algorithm?

### ISSUE THREAD

**Issue by touqir14** [google/jax#1467]
It is known that adaptive stochastic optimizers like Adam, Adagrad, RMSprop can indeed fail to converge. Moreover the test error of the trained models can be larger when compared to the models trained by SGD even though if they attain lower training error than with SGD - in other words they can overfit. See http://papers.nips.cc/paper/7003-the-marginal-value-of-adaptive-gradient-methods-in-machine-learning.pdf .

Recently, ADABOUND was proposed (https://openreview.net/pdf?id=Bkg3g2R9FX) to overcome the above mentioned problems of the adaptive optimizers. The algorithm combines the best of both worlds : (a) Makes fast progress initially like the adaptive methods, and (b) attains similar or better test accuracy to that of SGD.

Would this algorithm be welcome in JAX? If so, I would be interested to implement it.

**Comment by hawkinsp** [comment]
Looks interesting!

We want to be careful about what we add into JAX core; part of what lets us move quickly is that we have a fairly small API surface to support.

I'd suggest implementing this out of tree for now, given the optimizer is quite new and we don't yet know whether it will become popular or not. Of course, if it does become popular, we should certainly add it!

That said, one of the joys of JAX is that optimizers are only a few lines of code. You can probably just branch the ADAM implementation and adapt it without too much trouble:
https://github.com/google/jax/blob/d161e1badb6259f033d8f097f6ded55b6701d023/jax/experimental/optimizers.py#L328

What do you think?

**Comment by touqir14** [comment]
Just making sure that I understood correctly, by "implementing out of tree" were you suggesting to add this as a separate sort of experimental branch or create a JAX implementation and keep it outside of the repo?

**Comment by hawkinsp** [comment]
Yes, precisely. I'd start by implementing the new optimizer in your own code base. I'd estimate it will only be 10-15 lines of code. Should Adabound turn out to be important to many people, we can add it to Jax itself. (For future readers: feel free to +1 this issue if that's you!)

We want to be careful about what we add, since it's much easier to add things than to remove them later.

**Comment by touqir14** [comment]
In case you guys decide to add the Adabound algorithm to Jax, I have an implementation of it in my github : https://github.com/touqir14/Adabound-Jax . Actually, I am generally interested in writing optimization algorithms. Let me know if you guys would like to see any optimization algorithm implemented in Jax . @hawkinsp
